PhpAmqpLib/Wire/IO/StreamIO.php

```diff
@@ -231,15 +233,21 @@ public function read($len)
             }
 
             if ($buffer === '') {
+                $read_now = microtime(true);
+                $t_read = round($read_now - $read_start);
+                if ($t_read > $this->read_write_timeout) {
+                    throw new AMQPTimeoutException('Too many read attempts detected in StreamIO');
```

Comment:
Unfortunately, we've found this change does not solve the problem. Consumers usually consider AMQPTimeoutException as a normal situation, and intercept it.

For example, when the consumer in RabbitMqBundle exits, it closes the channel (https://github.com/php-amqplib/RabbitMqBundle/blob/master/RabbitMq/BaseAmqp.php#L77), and the channel calls `wait` without any timeout (https://github.com/php-amqplib/php-amqplib/blob/master/PhpAmqpLib/Channel/AMQPChannel.php#L169). Finally the consumer hangs at this point.

Actually, we've encountered the problem with closing channel in our test repo. But we've not realized the cause and removed the shutdown function. You can try to run different versions with the original consumer: https://github.com/maestroprog/phpamqplib-looping/commits/with-channel-closing

In our production environment, there is AMQPRuntimeException thrown (not AMQPTimeoutException) at this line. Consumers do not catch AMQPRuntimeException, they just crush. After that systemd restarts consumers. This workaround is acceptable for us because we have no consumers with timeout=0. But there should be no exception thrown from the library code when timeout=0 since the loop we try to avoid is a normal behavior for timeout=0.

I'd suggest to replace AMQPTimeoutException to AMQPRuntimeException in both StreamIO and AMQPReader, but I don't know if it is consistent. And maybe `$this->read_write_timeout` can be too small for throwing AMQPRuntimeException.

Reply:
> Consumers usually consider AMQPTimeoutException as a normal situation, and intercept it.

With this fix, that exception only reaches a consumer when `timeout` is greater than zero, which is what I would expect to happen.

> the channel calls `wait` without any timeout 

That's a separate issue to fix.

> $this->read_write_timeout can be too small for throwing AMQPRuntimeException.

It's up to the user to specify a good timeout value, and to be honest in the future (version `3.0.0`) an infinite timeout will probably be disallowed. I prefer to use `read_write_timeout` here instead of having a separate value.

I believe this PR fixes the issue where a `read` can wait indefinitely in StreamIO. Other issues will have to be dealt with on their own.

Reply:
Ok, I agree.